cli/internal/runsummary/repo_summary.go

```diff
@@ -0,0 +1,40 @@
+package runsummary
+
+import (
+	"github.com/vercel/turbo/cli/internal/ci"
+	"github.com/vercel/turbo/cli/internal/env"
+	"github.com/vercel/turbo/cli/internal/scm"
+	"github.com/vercel/turbo/cli/internal/turbopath"
+)
+
+type gitState struct {
+	Sha    string `json:"sha"`
+	Branch string `json:"branch"`
+}
+
+// getGitState returns the sha and branch when in a git repo
+// Otherwise it should return empty strings right now.
+// We my add handling of other scms and non-git tracking in the future.
+func getGitState(dir turbopath.AbsoluteSystemPath) *gitState {
+	allEnvVars := env.GetEnvMap()
+
+	gitstate := &gitState{}
+
+	// If we're in CI, try to get the values we need from environment variables
+	if ci.IsCi() {
+		vendor := ci.Info()
+		gitstate.Sha = allEnvVars[vendor.ShaEnvVar]
+		gitstate.Branch = allEnvVars[vendor.BranchEnvVar]
+	}
+
+	dirString := dir.ToString()
+	// Otherwise fallback to using `git`
+	if gitstate.Branch == "" {
+		gitstate.Branch = scm.GetCurrentBranch(dirString)
+	}
+	if gitstate.Sha == "" {
+		gitstate.Sha = scm.GetCurrentSha(dirString)
+	}
```

Comment:
I guess it is theoretically possible that we:

1. match CI,
2. these values are `""`
3. we ask `git` and it gives us a different answer.

But, since these are used to drive the actual CI behavior I don't find that likely. I think we should probably wrap this in an `else`.

```suggestion
	var vendor Vendor;

	if ci.IsCi() {
		vendor := ci.Info()
	}

	if vendor != nil {
		gitstate.Sha = allEnvVars[vendor.ShaEnvVar]
		gitstate.Branch = allEnvVars[vendor.BranchEnvVar]
	} else {
		// Otherwise fallback to using `git`
		gitstate.Branch = scm.GetCurrentBranch(dirString)
		gitstate.Sha = scm.GetCurrentSha(dirString)		
	}
```

(The question is "which is likely to produce a worse outcome in a failure mode" and this is my hypothesis.)

Reply:
I made the fallback in CI for a couple reasons

- our vendor definitions may have typos or incomplete (it is incomplete now)
- local runs which are trying to emulate a CI system but only partially setting the env vars

I agree the failure mode of env vars being set to `""` in a real CI system is unlikely.

I can still change this if you feel strongly about it, but I think I'm still leaning towards leaving this as is

Reply:
I don't have a strong enough opinion here to push it either direction. This can go either way.